Number_types/include/CGAL/FPU.h

```diff
@@ -347,16 +346,32 @@ inline double IA_bug_sqrt(double d)
 // With GCC, we can do slightly better : test with __builtin_constant_p()
 // that both arguments are constant before stopping one of them.
 // Use inline functions instead ?
-#define CGAL_IA_ADD(a,b) CGAL_IA_FORCE_TO_DOUBLE((a)+CGAL_IA_STOP_CPROP(b))
-#define CGAL_IA_SUB(a,b) CGAL_IA_FORCE_TO_DOUBLE(CGAL_IA_STOP_CPROP(a)-(b))
-#define CGAL_IA_MUL(a,b) CGAL_IA_FORCE_TO_DOUBLE(CGAL_IA_STOP_CPROP(a)*CGAL_IA_STOP_CPROP(b))
-#define CGAL_IA_DIV(a,b) CGAL_IA_FORCE_TO_DOUBLE(CGAL_IA_STOP_CPROP(a)/CGAL_IA_STOP_CPROP(b))
+#ifdef CGAL_ALWAYS_ROUND_TO_NEAREST
```

Comment:
IMO, if we are adding an extra case, we should simply assume that we do not know what the rounding mode is and not rely on to-nearest. I'm in favor of something like `CGAL_NO_FPU_ROUNDING_MODE_CHANGE`

Reply:
I guess, this implies more changes but that should be the way to go.

Reply:
It's an interesting point.

I think it's difficult to predict correct behavior with all existing and future rounding modes.

So I'd avoid a claim like CGAL_NO_FPU_ROUNDING_MODE_CHANGE.

But something like CGAL_CONSERVATIVE_INTERVALS might describe what's actually happening in this change, and then you could decide if that covers whatever floating point modes you care about or not.

I also notice that there are systems other than interval which care about floating point mode, and it's not clear to me that there will be a generic fix for all of them.

In which case CGAL_ALWAYS_ROUND_TO_NEAREST would be the right kind of flag to tell them about the particular environment they're operating in -- maybe some would need a CGAL_ALWAYS_ROUND_UPWARD, etc

Reply:
I don't think so. I think as long as you use `nextafter()` for the upper and -lower bound of the interval you are safe side no matter how the value was rounded

Reply:
For intervals, certainly -- but CGAL_NO_FPU_ROUNDING_MODE_CHANGE seems to make a more universal claim that everything in CGAL will always work without changing rounding mode.

Reply:
AFAIK, this is the only place where is is needed.

Reply:
Static filters do require round-to-nearest AFAIK (the bounds would be worse with arbitrary rounding), not sure if that's what you are talking about.

Reply:
Hmm, I see a number of cases that aren't clear to me.

Bounding_volumes/include/CGAL/Approximate_min_ellipsoid_d.h -- toward-zero and upward
Apollonius_graph_2/include/CGAL/Apollonius_graph_2/compare_quadratic.h -- upward
Number_types/include/CGAL/leda_bigfloat.h -- upward
Mesh_3/include/CGAL/Mesh_3/Robust_intersection_traits_3.h -- to nearest
Skin_surface_3/include/CGAL/Skin_surface_base_3.h -- to nearest
Straight_skeleton_2/include/CGAL/Straight_skeleton_2/Straight_skeleton_builder_traits_2_aux.h -- to nearest
Polynomial/test/Polynomial/Polynomial_traits_d_Residue.cpp -- upward
Polynomial/test/Polynomial/Polynomial_using_core.cpp -- upward
Polynomial/test/Polynomial/resultant.cpp -- upward
Polynomial/test/Polynomial/Polynomial_using_leda.cpp -- upward
Convex_hull_3/include/CGAL/convex_hull_3.h

Some of these don't seem to be part of filters, e.g., Approximate_min_ellipsoid.

My suggestion is to require explicit opt-in to any forced rounding mode.

i.e. If we use CGAL_ALWAYS_ROUND_TO_NEAREST then we can have CGAL_IA_SETFPCW throw an exception (or a compile-time error) when something tries to use a different mode, which will require systems to explicitly opt into CGAL_ALWAYS_AROUND_TO_NEAREST.

This should minimize surprising behavior, and allow gradual adoption, and allow alternate implementation for particular modes.

On the other hand, CGAL_NO_FPU_ROUNDING_MODE_CHANGE seems like it might require code to dynamically dispatch on the current rounding mode in the worst case, since it would need to operate in any existing or future rounding mode.

In any case, I'm operating from a position of ignorance, so maybe I'm being overly conservative -- but please make sure you're confident that you've considered all of the possibilities. :)